### ISSUE TITLE
Issue with switching windows in fullscreen at lower resolutions.

### ISSUE THREAD

**Issue by retep998** [SFML/SFML#306]
If I create a fullscreen window at a small resolution, such as 800x600, then I alt+tab, the fullscreen window remains on-screen and my resolution is still 800x600. Desired behavior is for the window to minimize and the resolution to restore to my desktop resolution when I switch to a different window. The following code can be used to demonstrate this issue.

``` C++
#include <SFML/Graphics.hpp>

int main() {
    sf::RenderWindow win(sf::VideoMode(800, 600, 32), "lol", sf::Style::Fullscreen);
    for (;;) {
        sf::Event e;
        while (win.pollEvent(e))
            if (e.type == sf::Event::KeyPressed)
                if (e.key.code == sf::Keyboard::Escape) std::exit(0);
        win.clear(sf::Color::Blue);
        win.display();
    }
}
```

Testing done with the latest git version of SFML using Visual Studio 2012 on Windows 8. Compiled for x64 in release mode.

**Comment by binary1248** [comment]
So... to get this discussion somewhere...

Would providing `minimize`/`maximize` methods in `sf::Window` be enough for the user to have something to work with?

On a `LostFocus` event (which is triggered on alt+tab), the user can recreate the window with the same resolution but in windowed mode and minimize it automatically with the new method.

On a `GainedFocus` event (which is triggered on alt+tab), the user can recreate the window with the same resolution but in fullscreen mode.

The `maximize` method is just there for symmetry reasons and to provide the user a way of... well... maximizing the window through code if their use case requires them to do so.

**Comment by MarioLiebisch** [comment]
Theoretically, although I have the feeling this should be handled by SFML. Having those methods would be nice to have in general. But I typically expect games to restore my desktop resolution, if I tab away.

**Comment by binary1248** [comment]
My solution restores the desktop resolution. When you recreate the window in windowed mode and instantly minimize it, it is as if you set the screen resolution back to desktop defaults directly. Having to go through the window to set the screen resolution is just something SFML specific but fits in rather well if you ask me. For the user, the behaviour is going to be the same as what you want. And with my proposal, we won't have to do much besides add those 2 methods (or 1 if we don't feel like adding maximize).

And about #755, it's not as simple as you think. I don't know if you know or not, but you don't have to completely destroy a window/context for it to lose access to its surface/resources. Because OpenGL ES is designed to run on memory-constrained systems, any time you switch activity the system is free to (doesn't _have to_) destroy all of your GL resources including the surface, which leads to the `EGL_BAD_SURFACE` error the reporter is getting. This is so that other activities that gain focus will have more to work with, which makes sense if you ask me. Simply transferring SFML's desktop context model over to mobile platforms just won't work.

**Comment by MarioLiebisch** [comment]
Yes, I'm aware it's more complicated than that, but I'm not sure you can recreate windows (never tried that, putting it on my todo). I'd just avoid instances where you have to exclude code paths for Android (or possibly iOS as well) just due to them handling windowing differently.

**Comment by binary1248** [comment]
I'm just saying the current implementation doesn't work properly already :wink:, so trying to preserve broken (for some users) behaviour doesn't make much sense to me. The fact is that recreating a window is guaranteed to work on desktop. If it doesn't work on mobile platforms, it has to be fixed, or a new model has to be designed with both platforms in mind. Saying that you are allowed to use the API in 1 way on desktop and in another way on mobile even though the interface stays the same isn't robust if you ask me.

**Comment by MarioLiebisch** [comment]
Well yeah, it's definitely an immediate solution, even if it's not perfect.

**Comment by Foaly** [comment]
Relates to #744.

**Comment by JonnyPtn** [comment]
Testing this on latest code, it still reproduces, but I believe it's expected behaviour for "exclusive fullscreen" and there's little we can do about it.

It's effectively a result of calling `ChangeDisplaySettings` which sets the display resolution to match the window - Without this you'd still get a window that behaves like a fullscreen one but only covering the window size area of the display.

I think if we _did_ want to solve this then we would need to keep track ourselves of the window size separately when in fullscreen and  scale the rendering appropriately to fill the actual display size, but I'm not sure if that is worth it, as for this scenario users should really be using borderless fullscreen
